### PR TITLE
JAX vector-valued QNode support using `host_callback.id_tap`

### PR DESCRIPTION
**Pull request by antalszava**
**Context**

Currently, the JAX interface doesn't support vector-valued QNodes.

Vector-valued QNodes are QNodes with return statements like:
* `qml.probs`, `qml.state` or `qml.sample` or;
* Multiple scalar return statements like `return qml.expval(0), qml.expval(1)` or;
* A mix of the aforementioned ones.

On top of that, the underlying JAX function (`host_callback.call`) only supports computing the gradient of scalar-valued functions (`jax.grad`).

**Changes**

Adds a new `pennylane/interfaces/batch/jax_jacobian.py` file that contains an implementation of the JAX interface using the `host_callback.id_tap` function instead of `host_callback.call`.

Users can pass the `jac_support` or `vector_valued_support` to switch to the `id_tap` version. Note: the `vector_valued_support` option is planned to be removed when #2034 allows vector-valued support by default.

**Benefits**

1. Computing the forward pass of vector-valued QNodes;
2. Computing the jacobian of vector-valued QNodes.

While 1. can be supported and is a work in progress in #2034, 2. is only feasible using `id_tap` as `host_callback.call` has a lack of support for `jax.jacobian`.

Therefore, it makes sense to have the two implementations live side-by-side. 

**Known unsupported cases:**

1. Using `jax.jit` on QNodes (https://github.com/google/jax/issues/9172);
2. QNodes with multiple probability returns with different number of wires specified;
3. Jacobian of QNodes in forward mode diff (e.g., with adjoint).

The lack of `jit` support likely is due to the side-effect nature of `id_tap`.

*Note for code review:* the `pennylane/interfaces/batch/jax.py` file was mostly left intact, although several changes are indicated by GitHub.

### CODE REVIEW

File: pennylane/interfaces/batch/jax.py [hunk 1]
```diff
@@ -1,299 +1,278 @@
...
-    tracing = any(isinstance(r, jax.interpreters.ad.JVPTracer) for r in res)
-
-    # When there are no tracers (not differentiating), we have the result of
-    # the forward pass and the jacobian, but only need the result of the
-    # forward pass
-    if len(res) == 2 and not tracing:
-        res = res[0]
-
-    return res
+# Copyright 2018-2021 Xanadu Quantum Technologies Inc.
```
Comment (antalszava) [inline pennylane/interfaces/batch/jax.py:1]: This file was renamed using `git mv`, that's why GitHub thinks there were major changes here.

Leaving a comment at each spot where there's a real change.

File: pennylane/interfaces/batch/jax.py [hunk 2]
```diff
@@ -1,299 +1,278 @@
...
+"""
+
+# pylint: disable=too-many-arguments
+import jax
+import jax.numpy as jnp
+from jax.experimental import host_callback
+
+import numpy as np
+import pennylane as qml
+from .jax_id_tap import _execute_id_tap, _execute_with_fwd_id_tap
```
Comment (antalszava) [inline pennylane/interfaces/batch/jax.py:26]: Added: import from `jax_id_tap`
Removed: `from pennylane.operation import Variance, Expectation`

File: pennylane/interfaces/batch/jax.py [hunk 3]
```diff
@@ -1,299 +1,278 @@
...
+        raise ValueError("The JAX interface only supports first order derivatives.")
+
+    for tape in tapes:
+        # set the trainable parameters
+        params = tape.get_parameters(trainable_only=False)
+        tape.trainable_params = qml.math.get_trainable_indices(params)
+
+    parameters = tuple(list(t.get_parameters()) for t in tapes)
+
+    vector_valued = gradient_kwargs.pop("jac_support", False)
```
Comment (antalszava) [inline pennylane/interfaces/batch/jax.py:70]: This line and subsequent lines using `vector_valued` or influenced by this object were added/updated.

File: pennylane/interfaces/batch/jax.py [hunk 4]
```diff
@@ -1,299 +1,278 @@
...
-        tapes=tapes,
-        device=device,
-        execute_fn=execute_fn,
-        gradient_fn=gradient_fn,
-        gradient_kwargs=gradient_kwargs,
-        _n=_n,
-    )
-
-
-def _validate_tapes(tapes):
```
Comment (antalszava) [inline pennylane/interfaces/batch/jax.py:90]: Validation and its calls were removed.

File: pennylane/interfaces/batch/jax.py [hunk 5]
```diff
@@ -1,299 +1,278 @@
...
-    tracing = any(isinstance(r, jax.interpreters.ad.JVPTracer) for r in res)
-
-    # When there are no tracers (not differentiating), we have the result of
-    # the forward pass and the jacobian, but only need the result of the
-    # forward pass
-    if len(res) == 2 and not tracing:
-        res = res[0]
-
-    return res
+# Copyright 2018-2022 Xanadu Quantum Technologies Inc.
```
Comment (antalszava) [inline pennylane/interfaces/batch/jax.py:1]: This file was renamed locally using `git mv`, hence GitHub shows major changes here. Leaving a comment at spots where a change occurred in reality.

**Comment by antalszava** [comment]
[sc-12935]

File: tests/interfaces/test_batch_jax.py [hunk 6]
```diff
@@ -610,3 +610,56 @@ def cost_fn(a, p, device):
...
+                [tape],
+                dev,
+                cache=cache,
+                interface='jax',
+                **execute_kwargs
+            )
+            return res[0]
+
+        res = jax.jacobian(cost)(params, cache=None)
+        assert res.shape == (1, 3)
```
[the remark this answers is not in the document]
Comment (antalszava) [reply tests/interfaces/test_batch_jax.py:636]: AFAIK no. :crying_cat_face:

File: pennylane/interfaces/batch/jax.py [hunk 7]
```diff
@@ -1,299 +1,286 @@
...
+
+    # Copy not to mutate the original dictionary if the same dictionary is
+    # being used for multiple executions
+    gradient_kwargs = deepcopy(gradient_kwargs)
+    jac_sup = gradient_kwargs.pop("jac_support", False)
+    vv_support = gradient_kwargs.pop("vector_valued_support", False)
+    vector_valued = jac_sup or vv_support
+
+    if gradient_fn is None:
+        exec_fwd = _execute_with_fwd if not vector_valued else _execute_with_fwd_id_tap
```
Comment (antalszava) [inline pennylane/interfaces/batch/jax.py:81]: These parts were extended such that the required options can be gathered.

**Comment by antalszava** [comment]
Closing this with #2110 merged. #1713 is a variant of #2110 using the experimental `id_tap` function. Once/if that function supports `jax.jit` fit to our use case, this approach could be reconsidered
